File: algorithms/Insertion Sort List/Insertion Sort List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var insertionSortList = function(head) {
    let ptr = head;
    
    while(ptr.next !== null){
        if(ptr.val <= ptr.next.val)
            ptr = ptr.next;
        else{
            let temp = ptr.next;
            ptr.next = ptr.next.next;
            
            
            if(temp.val < head.val)
                {
                    temp.next = head;
                    head = temp;
                }
            
            else{
                let ptr2 = head;
                while(ptr2.next != null  && temp.val >= ptr2.next.val){
                    ptr2 = ptr2.next;
                }
                temp.next = ptr2.next;
                ptr2.next = temp;
            }
        }
    }
    return head;
};

File: algorithms/Insertion Sort List/Insertion Sort List.java
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode cur = head;
        ListNode temp = new ListNode(-5001);
        ListNode prev = temp;
        while(cur != null){
            ListNode nxt = cur.next;
            if(prev.val >= cur.val)
                prev = temp;
            while(prev.next != null && prev.next.val < cur.val)
                prev = prev.next;
            cur.next = prev.next;
            prev.next = cur;
            cur = nxt;
        }
        return temp.next;
    }
}
